### ISSUE TITLE
[Admin][Reports] Missing products in Bulk co-op Packing Sheets report 

### ISSUE THREAD

**Issue by audez** [openfoodfoundation/openfoodnetwork#9844]
## Description
Some bulk products that have been ordered are not displayed in the Bulk co-op Packing Sheets report. 
And the "sum total" displayed is incorrect.

## Expected Behavior
There must be a line for each product, with the "sum total" for each product.

## Steps to Reproduce
1. Create 2 bulk products (products with the "group by option" set to yes)
2. Pass an order with these 2 products
3. Go to admin/reports/bulk_coop, choose "Bulk co-op Packing Sheets" report and see

You can also use shop 2735 on https://staging.coopcircuits.fr/

## Client1 ordered 1kg cacahuetes AND 1kg almonds
![Screen Shot 2022-10-23 at 12 59 16](https://user-images.githubusercontent.com/28386648/197388729-a9a021d8-9b65-447f-8a18-b283eaf64e4e.png)

## Only almonds appear in the Bulk co-op Packing Sheets report 
![Screen Shot 2022-10-23 at 12 58 42](https://user-images.githubusercontent.com/28386648/197388779-2482ea63-ac84-4c34-8a60-507ee4ebd145.png)

## Severity
bug-s4: it's annoying, but you can use it

**Comment by abdellani** [comment]
Hi 
I'll work on this

**Comment by abdellani** [comment]
Hi

The problem with this report is that we are grouping the [line items by order](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/reporting/reports/bulk_coop/packing_sheets.rb#L8
), then we pick [the first line item to fetch the data](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/reporting/reports/bulk_coop/base.rb#L79) needed to generate the report.

One easy way to solve the problem is to wrap every line item inside an array when `query_result` is called.
```Ruby
        def query_result
          table_items.map{|item| [item]}
        end
```

Another solution is to remove `group_by(:order)` from `query_result`, and to reimplement the needed methods from `Reporting::Reports::BulkCoop::Base` in `Reporting::Reports::BulkCoop::PackingSheets`. The new methods will expect one line item instead of multiple line items
For example:
```ruby
        def order_billing_address_name(line_items)
          billing_address = line_items.first.order.bill_address
          "#{billing_address.firstname} #{billing_address.lastname}"
        end
```
will become
```ruby
        def order_billing_address_name(line_item)
          billing_address = line_item.order.bill_address
          "#{billing_address.firstname} #{billing_address.lastname}"
        end
```

What do you think?

**Comment by jibees** [comment]
> The new methods will expect one line item instead of multiple line items

I think it's really a good move.

**Comment by abdellani** [comment]
Ok, I'll go with the second approach : )

**Comment by abdellani** [comment]
@jibees 
Now that I start testing the suggested solution, I think that not grouping the line items is not the solution.  
If two orders have the same product, the report will have two rows referring to the same product.

I tried to group by product.
=> if a product has multiple variants V1 and V2, the report will only show one variant V1 or V2.

I also tried to group by variants,
=> if two customers C1 and C2 order the same variant, the report will show only one customer C1 or C2.

I tried to group by variants and billing addresses [billing_address.id, variant.id], 
=> I noticed that if two orders are achieved with the same billing address, we will have two records representing the same billing address.
=> this means if the customer C1 makes two orders of the product P1, we'll have the two rows

The last attempt is a closer solution, I just needed to define a hash function that will return for two billing address that refers to the same customer a unique value. 
So we can group [billing_address_hash, variant.id]
```ruby
        def billing_address_hash(billing_address)
          exclude_list = ["id", "created_at", "updated_at"]
          billing_address.attributes.reject do |key|
            exclude_list.include?(key)
          end.hash
        end
```

What do you think?

**Comment by jibees** [comment]
> If two orders have the same product, the report will have two rows referring to the same product.

Right. I missed that. 

I'm wondering what exactly is the requirement here? 

Premises:
 * Shopper S_A checkouts Product P_1 variant V_1 quantity: 1
 * Shopper S_A checkouts Product P_1 variant V_2 quantity: 2
 * Shopper S_A re-checkouts Product P_1 variant V_2 quantity: 2
 * Shopper S_B checkouts Product P_1 variant V_1 quantity: 3

The report should be then:

| CUSTOMER | PRODUCT | VARIANT | SUM TOTAL |
|---|---|---|---|
| S_A | P_1 | V_1 | 1 |
| S_A | P_1 | V_2 | 4 (2+2) |
| S_B | P_1 | V_2 | 3 |


So basically, we want to group by Shopper, Product, Variant right?


Then, when not grouping them directly in query_result: 

##### `lib/reporting/reports/bulk_coop/packing_sheets.rb`
```ruby
        def query_result
          table_items.group_by do |line_item|
            [line_item.order.user, line_item.variant]
          end.values
        end
```


What am I missing? What do you think? Thanks!

**Comment by abdellani** [comment]
@jibees 
Yes, I'm grouping in `query_result` as you said.
The only difference is that I suggested using` [billing_address_hash(item.order.bill_address), item.variant.id]`
```ruby
        def query_result
          table_items.group_by do |item|
            [billing_address_hash(item.order.bill_address), item.variant.id]            
          end.values
        end
```

if we use `[line_item.order.user, line_item.variant]`, all customers :
1. who made an order as a guest during the checkout and, 
2. selected the same variant

 will be grouped in one row. 
=>  The reason is that for those orders, we have `order.user == nil`


also, the `customer` column in the report is calculated based on the billing address. I don't know why `order.customer` is not used since the Customer model contains `first_name` and `last_name`.

https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/reporting/reports/bulk_coop/packing_sheets.rb#L13
https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/reporting/reports/bulk_coop/base.rb#L79

Maybe we can use [line_item.order.customer, item.variant], I don't have any order for which `customer == nil`  except the first one that comes from the seed file.

What do you think?

**Comment by jibees** [comment]
Yep sorry, you're right, I think I meant the `customer` not the user:

```[line_item.order.customer, item.variant]```

Like customer: `# It's a relationship between a user and an enterprise but for guest orders it can also be between an email address and an enterprise.`
